docs: further clarify and reinforce the performance impacts of single page apps (SPAs) (#11637)
sveltejs/kit: documentation/docs/40-best-practices/05-performance.md

### Preventing waterfalls

> One of the biggest performance killers is what is referred to as a _waterfall_, which is a series of requests that is made sequentially. This can happen on the server or in the browser, but is especially costly when dealing with data that has to travel further or across slower networks such as a mobile user making a call to far away server.

Reviewer: ```suggestion
One of the biggest performance killers is what is referred to as a _waterfall_, which is a series of requests that is made sequentially. This can happen on the server or in the browser, but is especially costly when dealing with data that has to travel further or across slower networks such as a mobile user making a call to a far away server.
```

In the browser, waterfalls can occur when your HTML kicks off request chains such as requesting JS which requests CSS which requests a background image and web font. SvelteKit will largely solve this class of problems for you by adding [`modulepreload`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/modulepreload) tags or headers, but you should view [the network tab in your devtools](#diagnosing-issues) to check whether additional resources need to be preloaded.
- Pay special attention to this if you use web [fonts](#optimizing-assets-fonts) since they need to be handled manually.

> - Enabling [single page app (SPA) mode](single-page-apps) will cause such waterfall. With SPA mode, an empty page is generated, which fetches JavaScript, which ultimately loads and renders the page. This results in an extra network roundtrip before a single pixel can be displayed.

Reviewer: ```suggestion
- Enabling [single page app (SPA) mode](single-page-apps) will cause such waterfalls. With SPA mode, an empty page is generated, which fetches JavaScript, which ultimately loads and renders the page. This results in an extra network roundtrip before a single pixel can be displayed.
```

Waterfalls can also occur on calls to the backend whether made from the browser or server. E.g. if a universal `load` function makes an API call to fetch the current user, then uses the details from that response to fetch a list of saved items, and then uses _that_ response to fetch the details for each item, the browser will end up making multiple sequential requests. This is deadly for performance, especially for users that are physically located far from your backend.

> - Avoid this issue by using [server `load` functions](/docs/load#universal-vs-server) to make requests to backend services that are dependencies from the server rather than from the browser. Note, however, that server `load` functions are also not immune to waterfalls (though they are much less costly since they rarely involve roundtrips with high latency). For example if you query a database to get the current user and then use that data to make a second query for a list of saved items, it will typically be more performant to issue a single query with a database join.

Reviewer: ```suggestion
- Avoid this issue by using [server `load` functions](/docs/load#universal-vs-server) to make requests to backend services that are dependencies from the server rather than from the browser. Note, however, that server `load` functions are also not immune to waterfalls (though they are much less costly since they rarely involve roundtrips with high latency). For example, if you query a database to get the current user and then use that data to make a second query for a list of saved items, it will typically be more performant to issue a single query with a database join.
```

- If a universal `load` function makes an API call to fetch the current user, then uses the details from that response to fetch a list of saved items, and then uses _that_ response to fetch the details for each item, the browser will end up making multiple sequential requests. This is deadly for performance, especially for users that are physically located far from your backend. Avoid this issue by using [server `load` functions](/docs/load#universal-vs-server) where possible.
- Server `load` functions are also not immune to waterfalls (though they are much less costly since they rarely involve roundtrips with high latency). For example if you query a database to get the current user and then use that data to make a second query for a list of saved items, it will typically be more performant to issue a single query with a database join.

> 
> As an example, enabling [single page app (SPA) mode](single-page-apps) may be the single worst thing you can do to harm performance. When you load a page which has been generated with server-side rendering (SSR) or prerendering, the browser can start rendering the content as soon as it receives the initial response. With SPA mode, an empty page is generated, which fetches JavaScript, which ultimately loads and renders the page. This results in an extra network roundtrip before a single pixel can be displayed, which is disastrous for performance — especially for mobile users.

Reviewer 2: I think we should tone this one down, too

```suggestion
- Enabling [single page app (SPA) mode](single-page-apps) is another example of a waterfall. When you load a page which has been generated with server-side rendering (SSR) or prerendering, the browser can start rendering the content as soon as it receives the initial response. With SPA mode, an empty page is generated, which fetches JavaScript, which ultimately loads and renders the page. This results in an extra network roundtrip before a single pixel can be displayed, which harms performance — especially for mobile users where the network may be slow and JS can't be parsed as fast due to slower hardware.
```
